parlai/core/worlds.py

```diff
@@ -569,8 +569,18 @@ def batch_observe(self, index, batch_actions, index_acting):
                 hasattr(a, 'batch_observe')):
             batch_observations = a.batch_observe(batch_observations)
         else:
+            # Set internal states from batched agent
+            for attname in a.internal_states:
+                if not hasattr(a, 'batch_{}'.format(attname)): continue
+                for w, att in zip(self.worlds, getattr(a, 'batch_{}'.format(attname))):
+                    setattr(w.get_agents()[index], attname, att)
+
             batch_observations = [w.get_agents()[index].observe(o) for w, o in zip(self.worlds, batch_observations)]
 
+            # Get internal states and save in batched agent
+            for attname in a.internal_states:
+                setattr(a, 'batch_{}'.format(attname), [getattr(w.get_agents()[index], attname) for w in self.worlds])
+
```

Comment:
should we skip this if the agent doesn't have batch_act or batch_observe? maybe that extra code isn't worth the speedup